### PR TITLE
Add support for C#, F# and Visual Basic

### LINKED ISSUE #999: C# support?
I know I've been reluctant to add any JITted language. But a friend asked me... so I'm raising this!

https://docs.microsoft.com/en-us/dotnet/core/linux-prerequisites?tabs=netcore2x

dotNet core should install...and maybe there's a way to coax it to generate object code?

### PR DESCRIPTION
**Pull request by hez2010**
This PR supersedes #2592 (since #2592 seems inactive for months). 
Builder PR: https://github.com/compiler-explorer/misc-builder/pull/20

Add C#, F# and Visual Basic support via .NET 6 crossgen2.

Fixes #999
Re: #1180

Compiler options:
- `--targetos`: operating system, available options are `windows`, `linux` and `osx`. Default value is `windows`.
- `--targetarch`: architecture, available options are `arm`, `armel`, `arm64`, `x86` and `x64`. Default value is `x64`.
- `--instruction-set`: instruction set for compilation. Default value is `sse2`.
  - The allowable values for the --instruction-set option are described in the table below. Each architecture has a different set of valid instruction sets, and multiple instruction sets may be specified by separating the instructions sets by a `,`. For example `avx2,bmi,lzcnt`
  - Available instruction sets for each architecture:
    - arm: N/A
    - armel: N/A
    - arm64: base, neon, aes, crc, dotprod, rdma, sha1, sha2, lse
    - x86: base, sse, sse2, sse3, ssse3, sse4.1, sse4.2, avx, avx2, aes, bmi, bmi2, fma, lzcnt, pclmul, popcnt, avxvnni
    - x64: base, sse, sse2, sse3, ssse3, sse4.1, sse4.2, avx, avx2, aes, bmi, bmi2, fma, lzcnt, pclmul, popcnt, avxvnni
- `--singlemethodtypename`: Type name for single method compilation
- `--singlemethodname`: Name for single method compilation
- `--singlemethodindex`: Index for single method compilation
- `--singlemethodgenericarg`: Generic arguments for single method compilation
- `--codegenopt`, `--codegen-options`: Code generation options passed to JIT
- `-O`, `--optimize`: Enable optimizations (default)
- `--Od`, `--optimize-disabled`: Disable optimizations
- `--Os`, `--optimize-size`: Optimize code for size
- `--Ot`, `--optimize-time`: Optimize code for speed

Screenshots:

C#

image

F#

image

Visual Basic

image

### CODE REVIEW

**Comment by hez2010** [comment]
Hello, can someone help review the PR?

File: etc/config/fsharp.defaults.properties [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+compilerType=fsharp
+
+compiler.fsharp.exe=/opt/compiler-explorer/dotnet-6.0/.dotnet/dotnet
+compiler.fsharp.name=.NET 6 (x64)
+
+compiler.fsharp.clrDir=/opt/compiler-explorer/dotnet-6.0/artifacts/bin/coreclr/Linux.x64.Checked
+compiler.fsharp.runtimeId=linux-x64
+compiler.fsharp.targetOS=linux
+compiler.fsharp.targetArch=x64
+compiler.fsharp.instructionSet=base,sse,sse2,sse3,ssse3,sse4.1,sse4.2,avx,avx2,aes,bmi,bmi2,fma,lzcnt,pclmul,popcnt,avxvnni
```
Comment (partouf) [inline etc/config/fsharp.defaults.properties:13]: This is not how the (already existing) instructionSet property works. You probably want to rename this one to something like supportedInstructionSets and use that over here https://github.com/compiler-explorer/compiler-explorer/pull/3168/files#diff-e84770ce2518a6f397e8c06c65be9979d042091415ecd20b6a133159bd2bff37R140

File: lib/compilers/vb.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            }
+
+            cleanedAsm = cleanedAsm.concat(line + '\n');
+        }
+
+        return cleanedAsm;
+    }
+
+    async runCrossgen2(execOptions, crossgen2Path, jitPath, publishPath, dllPath, outputPath) {
+        const targetOS = this.compilerProps('compiler.vb.targetOS');
```
Comment (partouf) [inline lib/compilers/vb.js:138]: There's a lot of overlap/copy/pasting here, perhaps it's better to make a base function of sorts that does this and uses 'vb'/'csharp'/'fsharp' as a compilerKeyId variable in the class or use `this.key()` which seems to be the same.

File: lib/compilers/vb.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    optionsForFilter() {
+        return this.compilerOptions;
+    }
+
+    getOutputFilename(dirPath) {
+        return path.join(dirPath, `output.s`);
+    }
+
+    cleanAsm(asm) {
```
Comment (partouf) [inline lib/compilers/vb.js:113]: this seems duplicated, can you make this a shared base class function or a shared util function of some kind

**Review by partouf — CHANGES_REQUESTED** [review]
Thanks for this PR.

We probably need to discuss the installation of the binaries. The building of these don't really happen in the infra repo, but in special builders. Perhaps this one is appropriate: https://github.com/compiler-explorer/misc-builder
The resulting binaries will get uploaded to our S3 and then we can install them from infra.

About this PR: the general sense I have is that you probably want to add a dotnetbase class of some sort in between these 3 classes, there's a lot of duplicate code that will make maintaining this a lot harder if it remains duplicated.

File: etc/config/fsharp.defaults.properties (hunk 1 above)
Comment (hez2010) [reply etc/config/fsharp.defaults.properties]: Done.

File: lib/compilers/vb.js (hunk 2 above)
Comment (hez2010) [reply lib/compilers/vb.js]: Dome. I'm using `this.lang.id` instead.

File: lib/compilers/vb.js (hunk 3 above)
Comment (hez2010) [reply lib/compilers/vb.js]: Done. I've moved all duplicated code to `dotnet.js`.

Comment (partouf) [reply lib/compilers/vb.js]: Excellent

File: lib/compilers/vb.js (hunk 2 above)
Comment (partouf) [reply lib/compilers/vb.js]: Cool, smart

File: etc/config/csharp.defaults.properties [hunk 4]
```diff
@@ -0,0 +1,17 @@
+compilers=csharp
```
Comment (hez2010) [inline etc/config/csharp.defaults.properties:1]: Generally, I want to make it a compiler group so compiler explorer can compile code against a variety of versions of dotnet, and make both `targetOS` and `targetArch` configurable via compiler options, but I'd like to make it in another PR.

Comment (partouf) [reply etc/config/csharp.defaults.properties:1]: That was something I wanted to address as well. In that respect the compilerProps() calls probably also needs some tweaking. We can do it after this PR definitely.

File: etc/config/csharp.defaults.properties [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+supportsBinary=false
+needsMulti=false
+compilerType=csharp
+
+compiler.csharp.exe=/opt/compiler-explorer/dotnet-6.0/.dotnet/dotnet
```
Comment (hez2010) [inline etc/config/csharp.defaults.properties:6]: As per https://github.com/compiler-explorer/misc-builder/commit/cb00784922c3083c54237d448391570b15ddc597, should I change this path from `dotnet-6.0` to `dotnet-v6.0.0`?

Comment (partouf) [reply etc/config/csharp.defaults.properties:6]: most likely

**Comment by hez2010** [comment]
I've added support for trunk version and configurable options.

File: etc/config/csharp.defaults.properties (hunk 4 above)
Comment (hez2010) [reply etc/config/csharp.defaults.properties]: Just got some free time so done in this PR.

**Comment by hez2010** [comment]
I've verified it on a new setup machine, and it works without any issue.

**Comment by hez2010** [comment]
I have a question. Where can I set popular compiler options?

**Comment by hez2010** [comment]
.NET 6.0.1 has been released so I've bumped the version of dotnet to 6.0.1.

**Comment by partouf** [comment]
Hey

I have built the library with your changes, and installed it.
To test if it's a good build, I have tried the following command, but it failed:
What does the 

```
/opt/compiler-explorer/dotnet-v6.0.0/bin/coreclr/Linux.x64.Checked/crossgen2/crossgen2 --version
```
```
A fatal error occurred. The required library libhostfxr.so could not be found.
If this is a self-contained application, that library should exist in [/opt/compiler-explorer/dotnet-v6.0.0/bin/coreclr/Linux.x64.Checked/crossgen2/].
If this is a framework-dependent application, install the runtime in the global location [/usr/share/dotnet] or use the DOTNET_ROOT environment variable to specify the runtime location or register the runtime location in [/etc/dotnet/install_location].

The .NET runtime can be found at:
  - https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=ubuntu.20.04-x64&apphost_version=6.0.0-rc.1.21430.1
```

Do we have to install the dotnet runtime separately here or can we supply it with the build somehow?

**Comment by hez2010** [comment]
We have to execute crossgen2 with dotnet runtime (which has been placed in 
dotnet directory).
To run crossgen2: `path/to/.dotnet/dotnet path/to/crossgen2/crossgen2.dll`. 
Note that although the filename of crossgen2 contains a `.dll`, it's platform-neutral.

**Comment by partouf** [comment]
Ah, I see it should be in the tar. But the installer doesn't really support multiple root folders in a tar. Would be better if .dotnet and bin were both under artifacts. But maybe I can adjust the installer otherwise.

**Comment by hez2010** [comment]
I can move the `.dotnet` directory into `artifacts`. Will send a PR soon.

**Comment by partouf** [comment]
this installation seems to work https://github.com/compiler-explorer/infra/pull/649

**Comment by hez2010** [comment]
Ping

**Comment by mattgodbolt** [comment]
Sorry about this @hez2010 - we'll try and work out how to move forward soon!

**Comment by hez2010** [comment]
This PR is ready for review.

**Comment by hez2010** [comment]
Hello, is there any update or what else should I do to get this PR merged?

I deployed a CE with this PR branch for testing purpose: https://sysums.club:1443/

**Comment by VincentZalzal** [comment]
Yay! As an experienced C++ developer, but a new C# developer, I'm really happy to see .NET languages finally make it into Compiler Explorer :) My thanks to everyone involved in this.

**Comment by partouf** [comment]
So we're getting this error. It is probably a result of nsjail, but will investigate further

```
Welcome to .NET 6.0!
---------------------
SDK Version: 6.0.100

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
System.IO.FileLoadException: Could not load file or assembly 'System.Threading.Tasks.Dataflow, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. An internal error occurred.
 (0x8007054F)
File name: 'System.Threading.Tasks.Dataflow, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at Microsoft.Build.Execution.BuildManager.BeginBuild(BuildParameters parameters)
   at Microsoft.Build.Execution.BuildManager.BeginBuild(BuildParameters parameters, IEnumerable`1 deferredBuildMessages)
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, Boolean isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, String[] inputResultsCaches, String outputResultsCache)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)
Unhandled exception: System.IO.FileLoadException: Could not load file or assembly 'System.Threading.Tasks.Dataflow, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. An internal error occurred.
 (0x8007054F)
File name: 'System.Threading.Tasks.Dataflow, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at Microsoft.Build.Execution.BuildManager.BeginBuild(BuildParameters parameters)
   at Microsoft.Build.Execution.BuildManager.BeginBuild(BuildParameters parameters, IEnumerable`1 deferredBuildMessages)
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, Boolean isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, String[] inputResultsCaches, String outputResultsCache)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)
   at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args)
   at Microsoft.DotNet.Cli.Utils.MSBuildForwardingAppWithoutLogging.ExecuteInProc(String[] arguments)
Compiler returned: 82
```

**Comment by EgorBo** [comment]
From my understanding the current algorithm is to use Release .NET 6.0 and then run crossgen from Main (.NET 7.0)? That indeed can cause this sort of issues

**Comment by EgorBo** [comment]
A better approach is to prebuild .NET like this:

```
./build.sh Clr+Libs -c Release -rc Checked
cd src/tests
./build.sh Checked generatelayoutonly
```
then 
```
CORE_ROOT=artifacts/tests/coreclr/windows.x64.Checked/Tests/Core_Root
./dotnet.sh $CORE_ROOT/crossgen2/crossgen2.dll <path-to-my-dll> -r:$CORE_ROOT/*.dll <other-args>
```
Also, you don't need to use slow `dotnet publish` for console app, `dotnet build -c Release` should be enough and fast

**Comment by partouf** [comment]
I'm I think almost at a solution.

There are a couple of issues that I'm resolving:
* [x] It writes files and folders to /tmp, those should be going elsewhere
* [x] It opens too many files, causing things to crash, but we can adjust nsjail for that
* [ ] It requires an Internet connection to download nuget packages for runtime things, I'm planning to place the necessary files somewhere so they don't have to be downloaded

**Comment by partouf** [comment]
So this is a working version, will have to weave that into the code so the envvariables get set. Not all are needed probably, will try to minimize the number of changes.

`nsjail --config etc/nsjail/execute.cfg --time_limit=21 --cwd /app --bindmount /tmp/csharptmp:/app --env=LD_LIBRARY_PATH=/opt/compiler-explorer/dotnet-v6.0.1/bin/coreclr/Linux.x64.Checked/lib,/opt/compiler-explorer/dotnet-v6.0.1/bin/coreclr/Linux.x64.Checked/lib32,/opt/compiler-explorer/dotnet-v6.0.1/bin/coreclr/Linux.x64.Checked/lib64 --env=PATH=/opt/compiler-explorer/node/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/compiler-explorer/node/bin --env=DOTNET_CLI_TELEMETRY_OPTOUT=true --env=DOTNET_SKIP_FIRST_TIME_EXPERIENCE=true --env=DOTNET_SHARED_STORE=/app/tmp --env=temp=/app/tmp --env=NUGET_PLUGINS_CACHE_PATH=/app/tmp --env=NUGET_HTTP_CACHE_PATH=/app/tmp --env=DOTNET_ROOT=/opt/compiler-explorer/dotnet-v6.0.1/bin/coreclr/Linux.x64.Checked/.dotnet --env=NUGET_PACKAGES=/opt/compiler-explorer/dotnet_nuget-v6.0.0/packages --env=HOME=/app -- /opt/compiler-explorer/dotnet-v6.0.1/bin/coreclr/Linux.x64.Checked/.dotnet/dotnet build -c Release --self-contained --runtime linux-x64 -v q --nologo`

**Comment by kunalspathak** [comment]
When will this light up on https://godbolt.org/?

**Comment by partouf** [comment]
I'm finishing up the details, maybe tonight

**Comment by partouf** [comment]
So this is now live for .net 6.0.1 C# F# and VB, thank you all so much, especially @hez2010 and @omern1 !

**Comment by EgorBo** [comment]
Love it! Thanks!!

I think we can shorten the developer-loop (time to get the codegen) to less than a second if we address https://github.com/compiler-explorer/compiler-explorer/pull/3168#issuecomment-1024951986

I'll file a PR if @hez2010 not working on it already

Also, worth noting that it's AOT(R2R) quility of codegen which is usually worse than the JIT (especially the one with Dynamic PGO)

**Comment by partouf** [comment]
> Love it! Thanks!!
> 
> I think we can shorten the developer-loop (time to get the codegen) to less than a second if we address [#3168 (comment)](https://github.com/compiler-explorer/compiler-explorer/pull/3168#issuecomment-1024951986)
> 
> I'll file a PR if @hez2010 not working on it already
> 
> Also, worth noting that it's AOT(R2R) quility of codegen which is usually worse than the JIT (especially the one with Dynamic PGO)

Thanks, I briefly looked at making it work but failed, it will require some more diving into this if you really want it.
For now I'd recommend leaving it as is for a few weeks.

**Comment by hez2010** [comment]
> I have a question. Where can I set popular compiler options?
> 
> I want to provide two sets of compiler options:
> 1. Target x64 and enable all available instruction sets: `--targetarch x64 --instruction-set base,sse,sse2,sse3,ssse3,sse4.1,sse4.2,avx,avx2,aes,bmi,bmi2,fma,lzcnt,pclmul,popcnt,avxvnni -Ot`
> 2. Target ARM64 and enable all available instruction sets: `--targetarch arm64 --instruction-set base,neon,aes,crc,dotprod,rdma,sha1,sha2,lse -Ot`

Can these also be added to popular options?

**Comment by kunalspathak** [comment]
Just a note (I can create a separate issue), I don't see compiler options that pops up. Is there a way to add it or how do they get populated? I can open a PR.

image

**Comment by partouf** [comment]
To get possible compiler arguments, it require us to use the same environment settings from the compilation, but I don't think that's currently possible.

So perhaps we can solve it with something that Microsoft has been asking for their MSVC compilers https://github.com/compiler-explorer/compiler-explorer/issues/3070 as well

**Comment by ShreyasJejurkar** [comment]
First, a fall, thank you for all the contributors for the C#. This is great! 

Not a serious issue, but just putting it here, there is a delay in getting the ASM code for the given C# code. It's not instant like what Sharplab (https://sharplab.io/) offers. Is there anything that sharplab does that CompileExplorer doesn't do (Yet)?
Tagging @ashmind here, in case, if he has some input on the same!

**Comment by RubenRBS** [comment]
There is a PR currently pending review https://github.com/compiler-explorer/misc-builder/pull/27 that aims to improve compilation speeds, but we are always open for further help in this :)

**Comment by partouf** [comment]
> There is a PR currently pending review [compiler-explorer/misc-builder#27](https://github.com/compiler-explorer/misc-builder/pull/27) that aims to improve compilation speeds, but we are always open for further help in this :)

Even then, your results may vary. We're not dedicated website for the dotnet languages, we have a lot of complexity built in for all our other features that may be part to blame. Not much we can do about that.

**Comment by ShreyasJejurkar** [comment]
> There is a PR currently pending review [compiler-explorer/misc-builder#27](https://github.com/compiler-explorer/misc-builder/pull/27) that aims to improve compilation speeds, but we are always open for further help in this :)

Ohh, great! I wasn't aware @EgorBo has already opened PR for this! I wonder do @ashmind also does the same thing for sharplab!?

**Comment by hez2010** [comment]
@ShreyasJejurkar It can definitely be improved, but I just haven't gotten enough time to do it. We are currently using a slow approach which utilizes `dotnet publish -c Release` for build process, instead we can just use `dotnet build -c Release` and specify all necessary env vars (eg. `CoreRoot`) which can shrink the build process down to less than 1sec. It can be done in further PR.

**Comment by ShreyasJejurkar** [comment]
Yes yes @hez2010 no worries. Am just putting it here in case missed. And also bringing people together here who has built similar tools in .NET space, so that everyone can get benifits and learnings!
